### PR TITLE
Full support for compressed (Gzip, Brotli, Zlib) API requests

### LINKED ISSUE #2802: Support gzip, brotli, and deflate in requests and responses
We would like Meilisearch to accept user queries that are `content-encoded` in `gzip`, `br`, and `deflate`. This way, users will be able to send their payloads directly in gzip, brotli, or deflate to the engine, and the engine should be able to receive it blindly.

We are using actix-web for our HTTP engine. There already is [a middleware that is made explicitly for that](https://docs.rs/actix-web/latest/actix_web/middleware/struct.Compress.html).

Another functionality that this middleware should support is compressing the output of the engine too. It is useful when requesting many documents or facets, for example.

We want this feature to be tested to ensure that nothing wrong can happen when a user sends compressed documents or wants to receive a compressed HTTP response. You can look [at the actix-web tests to start](https://github.com/actix/actix-web/blob/ad7e67f940f0ebd25e0ad2bf1642b65b92df8e18/actix-web/src/middleware/compress.rs#L292-L302).

---

Edit from @curquiza 

## Impacted team

@meilisearch/docs-team this is something we would like to see mentioned in the documentation but NOT detailed. I mean, people should find the information about "the engine can accept gzip content" when typing "gzip" or "compressed". No need to detailed, this is something people already know if they need it, and 90% of the job is done by the browser

## TODO

- [ ] Implement the changes and merge to the `main` branch
- [ ] Update the related spec

### PR DESCRIPTION
**Pull request by mou**
# Pull Request

## Related issue
Fixes #2802 

## What does this PR do?
- Adds missed content-encoding support for streamed requests (documents)
- Adds additional tests to validate content-encoding support is in place
- Adds new tests to validate content-encoding support for previously existing code (built-in actix functionality for unmarshaling JSON payloads)

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

Thank you so much for contributing to Meilisearch!

### CODE REVIEW

**Comment by mou** [comment]
This is my first rust code in five years, so I am open to any suggestions for improvements. I did my best to separate the work I have done into commits that could be reviewed independently. I'm participating in Hacktoberfest ;)

**Comment by mou** [comment]
Unfortunately, to be able to reuse builtin actix logic and not copy-paste copyrighted code, I added additional dependency on `actix-http`

File: meilisearch-http/tests/index/create_index.rs [hunk 1]
```diff
@@ -18,6 +19,57 @@ async fn create_index_no_primary_key() {
     assert_eq!(response["details"]["primaryKey"], Value::Null);
 }
 
+#[actix_rt::test]
+async fn create_index_with_gzip_encoded_request() {
```
Comment (Kerollmops) [inline meilisearch-http/tests/index/create_index.rs:23]: Does it already support sending encoded responses when the accept-encoding header is set?
If not, could you please add one or two tests, please?

Comment (mou) [reply meilisearch-http/tests/index/create_index.rs:27]: I added two tests for response encoding. Had difficulties with `decode` function. I believe my implementation is not idiomatic (too much copying), so I will really appreciate your suggestions.

File: meilisearch-http/tests/index/create_index.rs [hunk 2]
```diff
@@ -36,6 +40,44 @@ async fn create_index_with_gzip_encoded_request() {
...
+        "uid": "test",
+        "primaryKey": None::<&str>,
+    }))
+    .unwrap();
+    let req = test::TestRequest::post()
+        .uri("/indexes")
+        .insert_header(Encoder::Gzip.header().unwrap())
+        .insert_header((ACCEPT_ENCODING, "br"))
+        .insert_header(ContentType::json())
+        .set_payload(Encoder::Gzip.encode(body))
```
Comment (Kerollmops) [inline meilisearch-http/tests/index/create_index.rs:65]: That's funny! If I understand correctly, you are accepting answers encoded as brotli but you sent your request as _GZIP, and it works_!

**Review by Kerollmops — CHANGES_REQUESTED** [review]
Would it be possible to fix [the clippy](https://github.com/meilisearch/meilisearch/actions/runs/3238445457/jobs/5312589295#step:5:717) and [rustfmt issues](https://github.com/meilisearch/meilisearch/actions/runs/3238445457/jobs/5312589012), please?

**Comment by mou** [comment]
> Would it be possible to fix [the clippy](https://github.com/meilisearch/meilisearch/actions/runs/3238445457/jobs/5312589295#step:5:717) and [rustfmt issues](https://github.com/meilisearch/meilisearch/actions/runs/3238445457/jobs/5312589012), please?

Sure. I run both of them before commit, but do not know why this errors was not reported (

File: meilisearch-http/tests/index/create_index.rs (hunk 2 above)
Comment (mou) [reply meilisearch-http/tests/index/create_index.rs:65]: Yes. I outgoing request encoded with Gzip, and I ask response to be encoded with Brotli.

**Comment by mou** [comment]
Sorry, once again. I should memorize "clippy, then rustfmt", not in reverse order )

**Comment by mou** [comment]
If you decide to support `zstd`, you only need to specify the corresponding feature for `actix` dependencies (and maybe write some tests to be sure it's enabled).

**Comment by gmourier** [comment]
Hey @mou and @Kerollmops 👋

I've tested it on my side it works like a charm! 

A little feedback, however, when I specify `Content-Encoding` while sending an un-compressed body I get this error. I think we can make the message field more explicit, WDYT?

```json
{
    "message": "Error that occur during reading payload: A payload reached EOF, but is not complete. Inner error: Some(Custom { kind: InvalidInput, error: \"invalid gzip header\" })",
    "code": "bad_request",
    "type": "invalid_request",
    "link": "https://docs.meilisearch.com/errors#bad_request"
}
```

**Comment by mou** [comment]
I did not touch error reporting at all. And I do not know how comprehensive this refactoring may become.

**Review by gmourier — APPROVED** [review]
Approving the feature usage, it's all good for me. Thanks, @mou!

**Review by Kerollmops — APPROVED** [review]
Thank you very much for your work. It is wonderful and will be merged as is 🎉
bors merge

**Comment by mou** [comment]
I hope it will save bandwidth on tasks like batch reindexing when the Internet separates index and source. Also, I believe it fits well into a SaaS solution built on top of meilisearch.

**Pull request merged**